packages/vats/src/core/basic-behaviors.js

```diff
@@ -53,22 +55,33 @@ const bootMsgEx = {
  * since we may want/need them later.
  */
 
-/** @param {BootstrapPowers} powers */
+/**
+ * @param {BootstrapPowers & {
+ *   produce: {vatStore: Producer<VatStore> }
+ * }} powers
+ *
+ * @typedef {{adminNode: any, root: unknown}} VatInfo as from createVatByName
+ * @typedef {MapStore<string, Promise<VatInfo>>} VatStore
+ */
 export const makeVatsFromBundles = ({
   vats,
   devices,
-  produce: { vatAdminSvc, loadVat },
+  produce: { vatAdminSvc, loadVat, vatStore },
 }) => {
   const svc = E(vats.vatAdmin).createVatAdminService(devices.vatAdmin);
   vatAdminSvc.resolve(svc);
-  // TODO: getVat? do we need to memoize this by name?
-  // TODO: rename loadVat to createVatByName?
+
+  /** @type {VatStore} */
+  const store = makeScalarMapStore();
+  vatStore.resolve(store);
+
   loadVat.resolve(bundleName => {
-    console.info(`createVatByName(${bundleName})`);
-    const root = E(svc)
-      .createVatByName(bundleName)
-      .then(r => r.root);
-    return root;
+    return provide(store, bundleName, _k => {
```

Comment:
I think this'll work, but I'm not excited about storing Promises long-term in Stores when we know it will be resolved reasonably soon (every time you send that Promise off-vat, swingset has to allocate a new `vpid` and then `syscall.resolve` it right away, which is a tiny bit ugly compared to just sending the Presence).  Also, if/when we make `bootstrap` Durable, this will have to change, since Durable Stores can't hold Promises.

I don't know of a decent way to benefit from the `provide` helper and also store the Presence. And open-coding `provide` would require extra work to get the handy memoization around `(bundleName)`. Maybe we could use an async form of `provide`, except it would have to return a Promise before resolution, and then a Presence afterwards, which would be a different kind of ugly.

Not suggesting you change it, just wanted to share my internal mental image of how Promises are used.

Reply:
> Maybe we could use an async form of `provide`,

I think I saw one of those around somewhere... I thought briefly about trying it...

> except it would have to return a Promise before resolution, and then a Presence afterwards, which would be a different kind of ugly.

I hadn't even thought of that. Yeah; I think I'll leave this as is.